tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc

```diff
@@ -0,0 +1,176 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h"
+#include <algorithm>
+#include <functional>
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+// Creates optimization profiles for a list of input shapes. The list of input
+// shapes are stored in shapes_.
+void TrtShapeOptimizationProfile::InitProfiles() {
+  if (input_shapes_.size() == 0) {
+    VLOG(1) << "Not creating profiles without input_shapes. "
+               "You have to enable profile generation mode first (build).";
+  } else {
+    VLOG(1) << "Creating profiles with startegy of one profile "
+            << "for each input (min=opt=max).";
+  }
+  for (auto& shape_vec : input_shapes_) {
+    std::vector<nvinfer1::Dims> dimvec;
+    for (auto& shape : shape_vec) {
+      dimvec.push_back(TensorShapeToTrtDims(shape, false));
+    }
+    // We set min=opt=max.
+    OptimizationProfileConfig profConfig{dimvec, dimvec, dimvec};
+    profiles_.push_back(std::move(profConfig));
+    VLOG(1) << "Created profile " << profiles_.back().DebugString();
+  }
+}
+
+#if IS_TRT_VERSION_GE(6, 0, 0, 0)
+Status TrtShapeOptimizationProfile::AddProfiles(
+    nvinfer1::IBuilder* builder, nvinfer1::IBuilderConfig* config,
+    const nvinfer1::INetworkDefinition* network) {
+  // Create a vector of optimization profiles
+  for (int i = 0; i < profiles_.size(); i++) {
+    auto* optProfile = builder->createOptimizationProfile();
+    Status status = profiles_[i].SetDimensions(network, optProfile);
+    if (!status.ok()) {
+      return status;
+    }
+    int idx = -1;
+    if (optProfile->isValid()) {
+      idx = config->addOptimizationProfile(optProfile);
+    }
+    if (idx >= 0) {
+      if (i != idx) {
+        return errors::Internal(
+            "Profile index of engine config is different from resource profile "
+            "index: ",
+            i, " != ", idx);
+      }
+      VLOG(1) << "Added optimization profile " << profiles_[i].DebugString()
+              << " to builder config.";
+    } else {
+      VLOG(ERROR) << "Failed to add optimization profile "
```

Comment:
Does VLOG(ERROR) even pass compilation? It should be LOG(ERROR) if that is what you want.